sdk/dapp-kit/src/components/AutoConnectWallet.tsx

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { ReactNode } from 'react';
+import { useEffect, useState } from 'react';
+import { useConnectWallet } from '../hooks/wallet/useConnectWallet.js';
+import { useWalletContext } from '../hooks/wallet/useWalletContext.js';
+import type { StorageAdapter } from '../utils/storageAdapters.js';
+import type { WalletAccount, WalletWithRequiredFeatures } from '@mysten/wallet-standard';
+
+type AutoConnectWalletProps = {
+	storageKey: string;
+	storageAdapter: StorageAdapter;
+	children: ReactNode;
+};
+
+export function AutoConnectWallet({
+	storageKey,
+	storageAdapter,
+	children,
+}: AutoConnectWalletProps) {
+	const { wallets, currentWallet: wallet, currentAccount: account } = useWalletContext();
+	const [previousWallet, setPreviousWallet] = useState<WalletWithRequiredFeatures | null>(null);
+	const [currentWallet, setCurrentWallet] = useState<WalletWithRequiredFeatures | null>(wallet);
+	const [previousAccount, setPreviousAccount] = useState<WalletAccount | null>(null);
+	const [currentAccount, setCurrentAccount] = useState<WalletAccount | null>(account);
+	const { mutate: connectWallet } = useConnectWallet();
+
+	// Instead of abstracting the previous wallet and account state logic into a generic
+	// usePrevious hook, we'll write this plainly to make it clear that this code depends
+	// on triggering shallow re-renders by updating state in the render code.
+	if (wallet !== currentWallet) {
+		setPreviousWallet(currentWallet);
+		setCurrentWallet(wallet);
+	}
+
+	if (account !== currentAccount) {
+		setPreviousAccount(currentAccount);
+		setCurrentAccount(account);
+	}
+
+	useEffect(() => {
+		if (wallet !== previousWallet || account !== previousAccount) {
```

Comment:
why are we comparing agains the previousWallet here? I feel like I am missing a bit of context on why we are tracking the previous state in the first place

Reply:
The "initial connection problem". If we model this entirely as effects, then we need to differentiate between these two events:
- The user disconnected their wallet (we should remove the stored state)
- The initial page load, when no wallet is initially connected (we should not remove the stored state)

Reply:
So if I am reading this right this might end up working something like:
* state 1: wallet: null, previousWallet: null -> no update
* state 2: wallet: wallet1, previousWallet: null -> update
* state 3: wallet: wallet2, previousWallet: null -> update
* state 4: wallet: wallet2, previousWallet: wallet1: -> update 
* state 5: wallet: null, previousWallet: wallet1: -> update
* state 6: wallet: null, previousWallet: wallet2: -> update

I might be wrong about if previousWallet gets updated before the effect runs when you call setPreviousWallet directly in render though, which would eliminate the extra effect calls

Reply:
I spoke with Jordan briefly about this whole implementation approach yesterday, and I think 
we're all on the same page that the event-based approach isn't super ideal and this approach isn't super ideal either because it makes the control flow harder to understand and adds unnecessary re-renders.

I think a more ideal solution might involve using Jotai for state management which makes handling things like this a lot simpler and is also better performance-wise. If it's okay with y'all, I'll land this PR as-is with some additional comments and a TODO to go back and improve this.